### ISSUE TITLE
Proposal: Enabling a pure-Python variant of Menpo

### ISSUE THREAD

**Issue by jabooth** [menpo/menpo#819]
### TL;DR 

The Menpo core library can provide a lot of value in many contexts outside of deformable modelling, but **we need to reduce the friction to use Menpo to make this practical**. We should enable a **pure-Python flavour of Menpo** that has no native components - installable as a **universal wheel on all platforms** with minimal dependencies.

### Menpo's opportunity

The Menpo core library can provide a lot of value in many contexts outside of deformable modelling. A core set of functionality is useful in most vision problems:

- Rich I/O functionality for images and video
- Great visualisation out of the box for 2D vision data
- Sophisticated transforms which operate equally on images (warping) and 2D/3D spatial data (transforms)
- Powerful generic abstractions and utilities: `LazyList`, `print_progress`
- Everything covered by over 500 unit tests 

None of this core functionality requires native extensions (Cython).

### The friction of native code

 Right now we have native code to provide the following features:

- Some image features (HOG, etc)
- Mesh normal computation
- Image patch extraction
- An (unused) native PWA implementation
- Faster path for affine image warping

Having native code adds a huge amount of friction to installing Menpo into existing Python environments. We have a sophisticated build system for Conda, but many potential users will not be able or willing to switch their entire Python ecosystem to use Menpo.

Furthermore, the native code adds a huge amount of overhead to maintaining Menpo for the devs:

- Maintaining binary builds of the full matrix of build targets (Linux/macOS/Windows, Python 2.7/3.5/3.6) adds a lot of complexity. 
- Whilst we learnt a lot setting all this stuff up, now many core Menpo developers are no longer working on Menpo full-time it's become a burden to keep all this stuff working 
- Just to add Python 3.6 support is a huge effort as we need to bump many granular packages independently. That's a shame as Menpo of course runs perfectly fine on Python 3.6.
- We could move to build wheels with our build system, but it's still going to be a significant on-going burden to build the full matrix of build targets, and Linux building of wheels is still challenging, so it would be an investment of time to just get this running.

Finally, I believe the native components make Menpo a lot less attractive to take on as a dependency for other projects. As an example the landmarker.io server is built on top of Menpo: it doesn't feel right that if you pip install a landmarking tool, native code is compiled for HOG feature extraction and mesh normal calculations. Other project maintainers I believe would feel the same.

### The trade off

If Menpo was pure-Python, we could:

- Make Menpo far more appealing as a 'glue' standard lib for Python
- Jettison a huge burden of maintenance from the core devs

and the downsides would be:

- some routines would get slower or become unavailable

I now feel that Menpo is being held back by the complexity of our native code, and I also don't feel it's essential to get a lot of utility out of menpo. People will not use Menpo in performance-sensitive environments, so I think **we are optimising for the wrong goal** (performance vs flexibility and ease of use). 

### The suggested path forward
I would like to:
-  Go through and either implement pure-Python variants of all our native features or make it so they are gracefully made unavailable if no native module is present
- Add a utility to quickly show users which native features are enabled or not so this doesn't become confusing
- Change the default `setup.py` experience to **not build native code by default** (Conda packages would still include native features)
- Upload a pure-Python wheel for Menpo that works on all platforms
- Change the website to advertise the pure-Python pip wheel variant of Menpo
- Allow users to build the native Menpo components optionally if they are willing to install Cython etc

I believe this would put Menpo in a much better position to remain relevant in the Python vision ecosystem going forward, and to remain maintainable.

**Comment by patricksnape** [comment]
This was completed by v0.10.0

Website was updated accordingly though we still advocate conda due to numpy scipy etc.
Menpo also exists on conda-forge so this isn't an issue
